### PR TITLE
Fakify params and weights under private config

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90417

Previously, we planned to lift the parameters and weights while exporting and implement our own transformer to "unlift" the lifted weights and params back to the graph as attributes. But this is bit challenging because:

- We need to maintain correct ordering for weights and parameters that are passed as inputs so that we know how to map them back. 
- Some weights are unused in the graph, so our transformer needs to be aware of which weights and parameters are not used in the graph. And we need to distinguish which are real user input and which are parameters. 
- There can be more edge cases we haven't seen in other models yet. 

I am aware that @Chillee  and @bdhirsh mentioned that functionalization won't work with fake-tensor attributes but this is fine for the short term as we don't expect users to be modifying weights and params in inference mode. In fact, we explicitly disable attribute mutation in torchdynamo export mode right now. 

Given above condition, it might be ok to just fakify params when we need. I use a flag to guard against this change.

Differential Revision: [D41891201](https://our.internmc.facebook.com/intern/diff/D41891201)

### CODE REVIEW

**Comment by zhxchen17** [comment]
Yeah I think this will be helpful to unblock all our use cases, including ASR and DPE models, since all models have weights. Right now it's nonscalable on our side to unlift params and buffers, because we have to handle so many corner cases.

**Comment by ezyang** [comment]
I think I've mentioned (to @eellison ? @voznesenskym ?) that I'm open to the idea of automatically fakeifying real tensors when in fake tensor mode in general, and then we would not have to guard this under a config flag. To understand if this is safe or not, we have to understand why we didn't add support for this initially.

Consider the following program:

```
x = torch.empty(0)
with FakeTensorMode():
    x.resize_(10)
    y = x * 2
```

What exactly is this program supposed to do? Naively, we would perform a resize_ on the original tensor x, but this is problematic, because the entire point of fake tensor mode is to avoid doing real compute, and a x.resize_(10) would cause real memory to be allocated. However, nor can we *not* do a resize_ on the original tensor x, because the subsequent computation on x needs the updated metadata from the resize call.

It sounds like you don't need this case. I think we'd be OK with throwing if this happens. Can we do this, instead of guarding this in private config?

File: torch/_subclasses/fake_tensor.py
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:29]: this should be a property of the FakeTensorMode, not a global property. Maybe `convert_non_fake_inputs`.

File: torch/_subclasses/fake_tensor.py
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:804]: as @ezyang commented, I think we need to keep track of the inputs we lift, and throw if we call torch.Tag.inplace_view` on them.

```
x = torch.rand([4, 4])
with FakeTensorMode(allow_non_fake_inputs=True):
     y = x.add_(3)
     y.resize_([16])
     assert(x.shape == y.shape) # would fail
```

File: torch/fx/experimental/proxy_tensor.py
Comment (eellison) [inline torch/fx/experimental/proxy_tensor.py:655]: constant_lifting is a bit of a misnomer at the moment. we don't have any guards in place to ensure that the arguments we lift are constant.

**Review by eellison — COMMENTED** [review]
I think checking for inplace views is sufficient here, but just thinking aloud, are there any other cases where someone would use `FakeTensorMode(allow_non_fake_inps=True)` and unwittingly affect real state ? Maybe by accessing the `param.grad` ?

**Comment by eellison** [comment]
We should also consider the api for if the user will want to make the lifted inputs static or dynamic

**Comment by tugsbayasgalan** [comment]
> I think checking for inplace views is sufficient here



> I think I've mentioned (to @eellison ? @voznesenskym ?) that I'm open to the idea of automatically fakeifying real tensors when in fake tensor mode in general, and then we would not have to guard this under a config flag. To understand if this is safe or not, we have to understand why we didn't add support for this initially.
> 
> Consider the following program:
> 
> ```
> x = torch.empty(0)
> with FakeTensorMode():
>     x.resize_(10)
>     y = x * 2
> ```
> 
> What exactly is this program supposed to do? Naively, we would perform a resize_ on the original tensor x, but this is problematic, because the entire point of fake tensor mode is to avoid doing real compute, and a x.resize_(10) would cause real memory to be allocated. However, nor can we _not_ do a resize_ on the original tensor x, because the subsequent computation on x needs the updated metadata from the resize call.
> 
> It sounds like you don't need this case. I think we'd be OK with throwing if this happens. Can we do this, instead of guarding this in private config?

Are you referring to mutation on global variable? I think local mutations on the inputs should be handled by functionalization right?

File: torch/fx/experimental/proxy_tensor.py
Comment (tugsbayasgalan) [reply torch/fx/experimental/proxy_tensor.py:655]: Sure I can rename and put this under FakeTensorMode

**Comment by ezyang** [comment]
Mutation on a global variable, but in this case it refers to parameters which you didn't fakeify.

File: torch/_subclasses/fake_tensor.py
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:795]: Can we update the error message now that this is supported and to suggest `allow_non_fake_inputs` ? And have a separate error msg for the `torch.Tag.inplace_view` case

**Comment by ezyang** [comment]
i'm going to let @eellison drive the review, holler if you need second opinion

File: torch/_subclasses/fake_tensor.py
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:728]: this is unused, plz delete. 
```
with FakeTensorMode(allow_non_fake_inputs=True):
     y = x.add_(3)
     print(isinstance(y, FakeTensor) # False
```

The reason we don't need it is because aten codegen returns the input instead of the output here.

File: torch/_subclasses/fake_tensor.py
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:793]: could we either `del flat_arg_tensors` after the call here or update it to be correct ? if someone were to add another use of `flat_arg_tensors` later it would be outdated

File: torch/_subclasses/fake_tensor.py
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:932]: update the error message to reflect the new flag here.

`f"Please convert all Tensors to FakeTensors first or instantiate FakeTensorMode with 'allow_non_fake_inputs'. 
 Found in {func}(*{args}, **{kwargs}) "`

File: torch/_subclasses/fake_tensor.py
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:929]: nit: maybe `Can't call metadata mutating ops on non-Fake Tensor Inputs.` ? inplace view is not a user-facing concept, and we might run with non globals/model attirbutes.

**Review by eellison — APPROVED** [review]
LGTM! just a few remaining small nits

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41891201).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41891201).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41891201).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41891201).

**Pull request closed without merging**
